libdrgn/program.c

```diff
@@ -1489,6 +1526,77 @@ drgn_thread_object(struct drgn_thread *thread, const struct drgn_object **ret)
 	return NULL;
 }
 
+static struct drgn_error *
+drgn_thread_name_linux_kernel(struct drgn_thread *thread, char **ret)
+{
+	struct drgn_error *err;
+	DRGN_OBJECT(comm, drgn_object_program(&thread->object));
+	err = drgn_object_member_dereference(&comm, &thread->object, "comm");
+	if (!err)
+		err = drgn_object_read_c_string(&comm, ret);
+	return err;
+}
+
+static struct drgn_error *
+drgn_thread_name_userspace_live(struct drgn_thread *thread, char **ret)
+{
+#define FORMAT "/proc/%" PRIu32 "/comm"
+	char path[sizeof(FORMAT)
+		- sizeof("%" PRIu32)
+		+ max_decimal_length(uint32_t)
+		+ 1];
+	snprintf(path, sizeof(path), FORMAT, thread->tid);
+#undef FORMAT
+	_cleanup_close_ int fd = open(path, O_RDONLY);
+	if (fd < 0)
+		return drgn_error_create_os("open", errno, path);
+	// Defined as 16 byte buffer in sched.h
+	// https://github.com/torvalds/linux/blob/master/include/linux/sched.h#L306
+	char buf[16];
```

Comment:
I missed this last time, but `/proc/$pid/comm` can actually be larger than 16 bytes (currently only for kernel threads, although there were discussions about allowing it more generally IIRC): https://github.com/torvalds/linux/blob/075dbe9f6e3c21596c5245826a4ee1f1c1676eb8/fs/proc/array.c#L101. Since this isn't tied to a binary file format, we might as well future-proof this and use at least 64.

Reply:
Oh wow, TIL!